### PR TITLE
Changed styling methods, regex keyword recognition, speedups and more

### PR DESCRIPTION
**Pull request by ludwig-austermann**
Ok after some time — here my implementation, where
- `strong-keywords` was removed in favour of `keyword-styles`
- `keyword-styles`, `main-text-styles` and `comment-styles` are now parameters accepting functions from `content -> content`
- `line-number-styles` is a parameter accepting a function `int -> content`
- `keywords` now also accepts items of type `regex` and have special behaviour depending on if non-word characters are in front or at the back of them
- some for loops with array pushing were changed to direct `map` calculation, making calculations more efficient
- `no-emph` was changed to `no-keyword`

I think the code is now better readable, except when having applied the loop conversions. Tests, examples and documentation was added/changed.

- [ ] In two places I commented some code out, because I think, that it is redundant — it should hence be removed before accepting.

### CODE REVIEW

File: algo.typ [hunk 1]
```diff
@@ -237,43 +231,35 @@
...
-        set text(..line-number-styles)
-        _numerals
-      },
+      line-number-styles(line-index),
       styles
     ).height
   )
 
-  // converting input parameters to absolute lengths
+  // TODO: WHY?!
```
Comment (ludwig-austermann) [inline algo.typ:261]: As far as I understand, taking the width directly has no different behaviour, or am I overseeing some weird special case?

File: algo.typ [hunk 2]
```diff
@@ -998,29 +993,27 @@
...
-  let block-inset-abs = measure(rect(width: block-inset), styles).width
-  let row-gutter-abs = measure(rect(width: row-gutter), styles).width
+  let indent-size = measure(
+    main-text-styles(raw("a" * tab-size)),
+    styles
+  ).width
 
-  let is-first-line = line-index == 0
-  let is-last-line = line-index == num-lines - 1
+  // TODO: WHY?!
```
Comment (platformer) [inline algo.typ:1013]: same comment as with `_algo-indent-guides`

File: test/test.typ [hunk 3]
```diff
@@ -620,7 +625,7 @@ def floyd_warshall(G):
 
 #code(
```
Comment (platformer) [inline test/test.typ:626]: This test seems to have failed. But if you can't figure it out, don't bang your head over it. The approach used for `code` is currently outdated, and this is a niche test case anyway. Yet another thing to update when improved tables arrive.

File: typst.toml [hunk 4]
```diff
@@ -7,3 +7,4 @@ license = "MIT"
 description = "Beautifully typeset algorithms."
 repository = "https://github.com/platformer/typst-algorithms"
 keywords = ["algorithm", "pseudocode", "code"]
+exclude = ["test", "examples/seperated"]
```
Comment (platformer) [inline typst.toml:10]: This line isn't actually necessary. I only ever upload the minimally necessary files when updating the package.

File: test/test.typ [hunk 5]
```diff
@@ -884,3 +889,104 @@ def floyd_warshall(G):
...
+
+#table(
+  columns: 2,
+  stroke: none,
+  align: (x, _) => (right, left).at(x),
+  "keywords:", `("A", "B\*", "\*C", "\*D\*", "\+", "E\*E", "FF?F", "G{4}")`,
+  "keyword-styles:", `x => text(red, x)`
+)
+
+#algo(
```
Comment (platformer) [inline test/test.typ:971]: I'd also like to test the simple case of a multi-word keyword with no special characters on the end, just to make sure it's working. i.e. if `keywords` includes `"hello world"`:

```typst
...
blah hello world blah // hello world is highlighted
blahhello world blah // no highlight
blah hello worldblah // no highlight
...
```

File: test/test.typ [hunk 6]
```diff
@@ -884,3 +889,104 @@ def floyd_warshall(G):
...
+  ```
+]
+
+== Algo Keyword Regex Test
+
+#table(
+  columns: 2,
+  stroke: none,
+  align: (x, _) => (right, left).at(x),
+  "keywords:", `("A", "B\*", "\*C", "\*D\*", "\+", "E\*E", "FF?F", "G{4}")`,
```
Comment (platformer) [inline test/test.typ:967]: Make sure to update this. Later I'll rewrite this document to use idwtet, but unfortunately it wasn't available at the time I was making this.

**Review by platformer — CHANGES_REQUESTED** [review]
The changes mostly look good, just some small things to address. I think making better use of array functions is a good idea.

**Pull request closed without merging**